icefoganalytics/internal-data-portal
# User: Dataset Description Read Page

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/44

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)

# Context

## Dataset Description Read Page - /datasets/:slug/read

Slug should be pre-computed from Dataset#name.

Based on User policy logic, there should be an Editor and Consumer view for these pages.

Note that these pages don’t have any save buttons, so are assumed to auto-save a few seconds after edits. The pages should probably have some kind of “saved” icon that enters a loading state after edits, then goes back to a green check mark or something.

The wireframe on page 4 might be a read only view for the minimum permission user, with the wireframe on page 8 being the edit view.

There seems to be different read views depending on user permission level as well.

### Front-end routes

* /datasets/:slug/description
* /datasets/:slug/description/manage
* /datasets/:slug/fields
* /datasets/:slug/fields/manage
* /datasets/:slug/access/manage - non-manage page does not exist
* /datasets/:slug/change-log
* /datasets/:slug/change-log/manage
* /datasets/:slug/visualize
* /datasets/:slug/visualize/manage

### Back-end routes

* /datasets/:id - GET or maybe /datasets/:slug_or_id
* /access-requests - POST, create new access request

### Components

* `BaseLayout.vue` - see previous
* `DatasetLayout.vue`
    * Title - Dataset#name
    * `DescriptionTab.vue` - loads DatasetDescriptionEditPage.vue or DatasetDescriptionReadPage.vue
    * `FieldsTab.vue` - loads DatasetFieldsPage.vue
    * `VisualizeTable.vue` - loads DatasetVisualizePage.vue
* `DatasetDescriptionReadPage.vue` - for minimum permission User
    * `RequestAccessDialog.vue`
        * Name - UserSelector.vue, AccessRequest#requestor_id
        * Department - might be pre-filled from requestor department? Or be independently selectable, in which case it needs to be tracked on the AccessRequest model.
        * Request on behalf of (Program/App) - AccessRequest#project_name
        * Project Description - AccessRequest#project_description
        * Cancel button
        * Request button - POST to /access-requests, create action
    * `DataDescriptionCard.vue`
        * `RequestAccessButton.vue` - opens RequestAccessDialog.vue
        * Name - Dataset#name
        * Description - Dataset#description
        * Spatial? - Dataset#is_spacial_data
        * Live data? - Dataset#is_live_data
        * Terms of Use - Dataset#terms_of_use
        * Credits - Dataset#credits
    * `OwnerCard.vue`
        * Owner Name - UserSelector.vue, Dataset#owner_id
        * Owner Position - pulled from owner position?
        * Support Name - UserSelector.vue, Dataset#supporting_user_id
        * Support Email - pulled from the supporting user?
        * Support Position - pulled from the supporting user?
        * Department - pulled from owner department?
        * Division - pulled from owner division?
        * Branch - pulled from owner branch?
        * Unit - pulled from owner unit?
    * `SubjectCard.vue`
        * Keywords - TagsCombobox.vue, multi select for tags probably needs a lot of very complex wiring to support mapping tags through taggings to dataset by id, and also supporting creation of new tags and linking them via taggings

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3b9a8065-cdc9-4968-897b-1c245c4884f5)

# Implementation

Add a class method to the dataset for returning the relevant access grant for a given user.
Add a scope to the dataset for filtering datasets by those accessible to a given user.
Scope dataset table view results based on user type and access grant accessibility.
Add dataset description read page with call-to-actions for subscribing and requesting access.
Also add dataset "access type badge" to datasets read page.

# Screenshots

Viewing dataset that is subscribable
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/f52680d6-38ed-47f7-8047-20378a8ce03d)

Viewing dataset with active subscription
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/947ba671-ec57-4c86-8a0c-5a6968bb6098)

Viewing dataset with open access
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/c997c887-13f1-44f1-937b-4e01e9e4d7cb)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080 and go to the /dashboard page.
4. Using the QA Scenario, switch to the Data Owner role.
5. Create 4 datasets.
6. First dataset should have an access grant with open access.
7. Second dataset should have an access grant with "Self Serve" access.
8. Third dataset should have an access grant with "Screened" access.
9. Fourth dataset should have no matching access grants.
10. Go back to the "DataPortal" breadcrumb (/dashboard).
11. Cycle through user types until you become a "user". _cycle user button will read `CYCLE USER TYPE TO DATA_OWNER`_.
12. View the existing datasets via the "View Datasets" button.
13. Now that you are a user, the datasets you can view will be restricted by the back-end policy scoping. You will only see Datasets that you can view, subscribe to, or request access to. _I haven't implemented the table action buttons yet._
14. Click on the open access one.
15. You can see the new dataset description "read" page.
16. Note that the top right corner has an info badge with describing your access, in this case "open access".
17. Go back to the Datasets page via the breadcrumb.
18. Click on the dataset with the "Self Serve" access type.
19. Note that in the top left you can see a "Subscribe" call to action button.
20. Click on it and fill in the required fields.
21. Note that the "Project Description" is required if you said it should be when creating the access grant.
22. After you submit the modal form, the page will refresh, and the top right button will convert to an info badge showing "Subscribed".
23. Go back to the Datasets page via the breadcrumb, again.
24. Note that the table entry now shows you as subscribed to the dataset you just subscribed to.
25. Click on the last dataset entry that shows access type as Screened access.
26. The call-to-action button in the top right corner shows "Request Access".
27. Complete the access request.
28. After the request completes the call-to-action button is replaced by a "Request Pending" info badge.
